### ISSUE TITLE
Better shared memory allocation under Docker

### ISSUE THREAD

**Issue by lw** [pytorch/pytorch#58386]
## 🚀 Feature
Use the `memfd_create` syscall (when available) instead of `shm_open` to allocate shared memory to evade the system limits on how much shared memory can be allocated.

## Motivation

From "empirical" experience, it seems that CPU tensors backed by shared memory are allocated as files inside `/dev/shm`. (Not sure if this happens "directly", through `open`, or indirectly, through `shm_open`). The problem is that `/dev/shm` has a limited capacity, and this is particularly pronounced inside Docker containers, which sets it by default to a few megabytes total. While Docker's config can be changed, many users don't know this. We've gotten reports about this over and over in [PyTorch-BigGraph](https://github.com/facebookresearch/PyTorch-BigGraph), which can sometimes allocate several GBs in shared memory.

Instead, we could use `memfd_create`, which allocates "regular" memory (not subject to those capacity limits), while still giving us a file descriptor that can be passed among processes thus allowing to effectively share it. The `memfd_create` syscall is rather new (~6-7 years ago), so we would still have to fallback to `shm_open` when it's not available.

Note: TensorPipe already uses `memfd_create` to allocate its shared-memory buffers for IPC.

**Comment by imaginary-person** [comment]
Do users usually encounter `SIGBUS` if they don't change the default Docker config? Just asking, so that someone searching for keywords of such an issue may land here.
Thanks!

**Comment by lw** [comment]
Yes, we tend to detect these issues as SIGBUS failures, which also manifest as "Bus error" messages or as error code -7 (to add a few more keywords).
